File: internal/game/states/armory.go
package states

import (
	"fmt"

	"github.com/slycrel/slycrel/internal/game"
	"github.com/slycrel/slycrel/internal/model"
)

// ArmoryState shows the armory ANSI menu.
// Original: ST_Armory
type ArmoryState struct{}

func (ArmoryState) ID() game.StateID { return "armory" }

func (ArmoryState) Enter(s *game.Session) {
	s.Character.Location = model.TheArmory
	s.IO.ShowANSIFile("armory_menu")
	s.SetNextState("armory_prompt")
}

// ArmoryPromptState shows the armory menu.
// Original: ST_ArmoryShortPrompt + ST_ArmoryMenu
type ArmoryPromptState struct{}

func (ArmoryPromptState) ID() game.StateID { return "armory_prompt" }

func (ArmoryPromptState) Enter(s *game.Session) {
	c := s.Character
	s.IO.Outln(fmt.Sprintf("  Coins on hand: %d", c.CoinsHand), true, 4)
	s.IO.Outln("[W]eapons, [A]rmor, [S]ell Weapon, [G] Sell Armor, [M]ain/Off Switch, [Q]uit, [?]Help", true, 1)
	s.IO.Cr()

	choice := s.IO.LettersPrompt("Your choice?", "WASGMQ?", 1, true, true)

	switch choice {
	case "W":
		s.SetNextState("buy_weapon")
	case "A":
		s.SetNextState("buy_armor")
	case "S":
		s.SetNextState("sell_weapon")
	case "G":
		s.SetNextState("sell_armor")
	case "M":
		// Swap weapon slots
		s.IO.Cr()
		s.IO.Outln("You switch your weapons.", true, 1)
		s.IO.Cr()
		c.Weapons[0], c.Weapons[1] = c.Weapons[1], c.Weapons[0]
		s.SetNextState("armory_prompt")
	case "Q":
		s.Character.Location = model.TheTown
		s.SetNextState("town")
	case "?":
		s.SetNextState("armory")
	default:
		s.SetNextState("armory_prompt")
	}
}

// BuyWeaponState lists and lets you buy weapons.
// Original: ST_OpenWeaponsFile + ST_ChooseWeapon + ST_QBuyWeapon + ST_QReplaceWeapon + ST_DisplayWeapons
type BuyWeaponState struct{}

func (BuyWeaponState) ID() game.StateID { return "buy_weapon" }

func (BuyWeaponState) Enter(s *game.Session) {
	weapons, err := s.Store.LoadWeapons()
	if err != nil || len(weapons) == 0 {
		s.IO.Outln("No weapons available!", true, 6)
		s.IO.PausePrompt("--press a key--")
		s.SetNextState("armory_prompt")
		return
	}

	s.IO.Cr()
	s.IO.Outln("=--=-- Available Weapons ---=--=", true, 3)
	s.IO.Cr()

	for i, w := range weapons {
		affordable := " "
		if int64(w.Cost) <= s.Character.CoinsHand {
			affordable = "*"
		}
		rangeStr := ""
		if w.Range > 0 {
			rangeStr = fmt.Sprintf(" Range:%d", w.Range)
		}
		s.IO.Outln(fmt.Sprintf(" %s%2d) %-18s Strike:%-3d%s  Cost:%-4d",
			affordable, i+1, w.Name, w.Strike, rangeStr, w.Cost), true, 1)
	}
	s.IO.Cr()
	s.IO.Outln("  (* = you can afford)", true, 4)
	s.IO.Cr()

	num := s.IO.NumbersPrompt("Choose Your Weapon (0=Quit):", 0, len(weapons))
	if num == 0 {
		s.SetNextState("armory_prompt")
		return
	}

	weapon := weapons[num-1]

	if !s.IO.YesNoQuestion(fmt.Sprintf("Buy %s for %d coins? [Y/n]", weapon.Name, weapon.Cost)) {
		s.IO.Outln("FINE, I didn't want your stupid weapon anyway...", true, 1)
		s.IO.Cr()
		s.SetNextState("armory_prompt")
		return
	}

	if s.Character.CoinsHand < int64(weapon.Cost) {
		s.IO.Cr()
		s.IO.Outln("Sorry, I'm not giving these away. Come back when you get more money.", true, 6)
		s.IO.Cr()
		s.IO.PausePrompt("--More--")
		s.SetNextState("armory_prompt")
		return
	}

	// Show current weapons and ask which slot
	s.IO.Cr()
	w1name := s.Character.Weapons[0].Name
	if w1name == "" {
		w1name = "Empty"
	}
	w2name := s.Character.Weapons[1].Name
	if w2name == "" {
		w2name = "Empty"
	}
	s.IO.Outln(fmt.Sprintf("  Weapon 1: %s", w1name), true, 1)
	s.IO.Outln(fmt.Sprintf("  Weapon 2: %s", w2name), true, 1)
	s.IO.Cr()

	slot := s.IO.NumbersPrompt("Which weapon slot to replace? (1 or 2):", 1, 2)

	s.Character.Weapons[slot-1] = weapon
	s.Character.CoinsHand -= int64(weapon.Cost)
	s.IO.Cr()
	s.IO.Outln(fmt.Sprintf("Okay, you buy the spiffy %s.", weapon.Name), true, 3)
	s.IO.PausePrompt("--More--")
	s.SetNextState("armory_prompt")
}

// BuyArmorState lists and lets you buy armor.
// Original: ST_OpenArmorFile + ST_ChooseArmor + ST_QBuyArmor + ST_SaveNewArmor
type BuyArmorState struct{}

func (BuyArmorState) ID() game.StateID { return "buy_armor" }

func (BuyArmorState) Enter(s *game.Session) {
	armors, err := s.Store.LoadArmor()
	if err != nil || len(armors) == 0 {
		s.IO.Outln("No armor available!", true, 6)
		s.IO.PausePrompt("--press a key--")
		s.SetNextState("armory_prompt")
		return
	}

	s.IO.Cr()
	s.IO.Outln("=--=-- Available Armor ---=--=", true, 3)
	s.IO.Cr()

	classNames := []string{"Body", "Shield", "Full Body"}
	for i, a := range armors {
		affordable := " "
		if int64(a.Cost) <= s.Character.CoinsHand {
			affordable = "*"
		}
		cn := "Body"
		if int(a.Class) < len(classNames) {
			cn = classNames[a.Class]
		}
		s.IO.Outln(fmt.Sprintf(" %s%2d) %-18s Def:%-3d Type:%-9s Cost:%-4d",
			affordable, i+1, a.Name, a.Defense, cn, a.Cost), true, 1)
	}
	s.IO.Cr()
	s.IO.Outln("  (* = you can afford)", true, 4)
	s.IO.Cr()

	num := s.IO.NumbersPrompt("Choose Your Armor (0=Quit):", 0, len(armors))
	if num == 0 {
		s.SetNextState("armory_prompt")
		return
	}

	armor := armors[num-1]

	if !s.IO.YesNoQuestion(fmt.Sprintf("Buy %s for %d coins? [Y/n]", armor.Name, armor.Cost)) {
		s.IO.Outln("Okay, your loss. Have fun dying.", true, 1)
		s.IO.Cr()
		s.SetNextState("armory_prompt")
		return
	}

	if s.Character.CoinsHand < int64(armor.Cost) {
		s.IO.Cr()
		s.IO.Outln("Sorry, I'm not giving these away. Come back when you get more money.", true, 6)
		s.IO.Cr()
		s.SetNextState("armory_prompt")
		return
	}

	// Replace body armor (slot 0)
	s.Character.Armor[0] = armor
	s.Character.CoinsHand -= int64(armor.Cost)
	s.IO.Cr()
	s.IO.Outln(fmt.Sprintf("Okay, you buy the spiffy %s.", armor.Name), true, 3)
	s.IO.PausePrompt("--More--")
	s.SetNextState("armory_prompt")
}

// SellWeaponState sells the primary weapon.
// Original: ST_QSellWeapon + ST_SaveWeaponSale
type SellWeaponState struct{}

func (SellWeaponState) ID() game.StateID { return "sell_weapon" }

func (SellWeaponState) Enter(s *game.Session) {
	w := s.Character.Weapons[0]
	sellPrice := w.Cost / 2

	if w.Name == "" || w.Name == "Hands" {
		s.IO.Cr()
		s.IO.Outln("You can't sell your hands. Idiot...", true, 6)
		s.IO.Cr()
		s.IO.PausePrompt("-More-")
		s.SetNextState("armory_prompt")
		return
	}

	if w.Cost <= 0 {
		s.IO.Cr()
		s.IO.Outln("Sorry, I don't buy junk.", true, 6)
		s.IO.Cr()
		s.IO.PausePrompt("-More-")
		s.SetNextState("armory_prompt")
		return
	}

	if !s.IO.YesNoQuestion(fmt.Sprintf("Sell your %s for %d coins? [Y/n]", w.Name, sellPrice)) {
		s.IO.Cr()
		s.IO.Outln("Okay, you jerk, I didn't want it anyway.", true, 3)
		s.IO.Cr()
		s.IO.PausePrompt("-More-")
		s.SetNextState("armory_prompt")
		return
	}

	s.Character.CoinsHand += int64(sellPrice)
	s.Character.Weapons[0] = model.Weapon{Name: "Hands", Strike: 1, ActionStr: "punch"}
	s.IO.Cr()
	s.IO.Outln("Okay, thanks. Come back anytime.", true, 1)
	s.IO.Cr()
	s.IO.PausePrompt("-More-")
	s.SetNextState("armory_prompt")
}

// SellArmorState sells the body armor.
// Original: ST_QSellArmor + ST_SaveArmorSale
type SellArmorState struct{}

func (SellArmorState) ID() game.StateID { return "sell_armor" }

func (SellArmorState) Enter(s *game.Session) {
	a := s.Character.Armor[0]
	sellPrice := a.Cost / 2

	if a.Name == "" || a.Cost <= 0 {
		s.IO.Cr()
		s.IO.Outln("Sorry, I don't buy junk.", true, 6)
		s.IO.Cr()
		s.IO.PausePrompt("-More-")
		s.SetNextState("armory_prompt")
		return
	}

	if !s.IO.YesNoQuestion(fmt.Sprintf("Sell your %s for %d coins? [Y/n]", a.Name, sellPrice)) {
		s.IO.Cr()
		s.IO.Outln("Okay, you jerk, I didn't want it anyway.", true, 3)
		s.IO.Cr()
		s.IO.PausePrompt("-More-")
		s.SetNextState("armory_prompt")
		return
	}

	s.Character.CoinsHand += int64(sellPrice)
	s.Character.Armor[0] = model.Armor{}
	s.IO.Cr()
	s.IO.Outln("Okay, thanks. Come back anytime.", true, 1)
	s.IO.Cr()
	s.IO.PausePrompt("-More-")
	s.SetNextState("armory_prompt")
}

File: clients/web/game/states/armory.js
import { WhereType } from '../../model/enums.js';
import { loadWeapons, loadArmor, saveCharacter } from '../../store/local.js';
import { TownState } from './town.js';

const ARMOR_CLASS_NAMES = ['Body', 'Shield', 'Full Body'];

// Armory — port of internal/game/states/armory.go.
// Loads weapons.json and armor.json on demand and lets the player buy,
// sell, or swap their primary/off-hand weapon slot.

export class ArmoryState {
  async enter(session) {
    session.character.location = WhereType.TheArmory;
    session.io.clear();
    await session.io.showAnsiFile('armory_menu');
    session.setNext(new ArmoryPromptState());
  }
}

export class ArmoryPromptState {
  async enter(session) {
    const c = session.character;
    const { io } = session;
    io.println(`  Coins on hand: ${c.coinsHand}`, 4);
    io.println('[W]eapons, [A]rmor, [S]ell Weapon, [G] Sell Armor, [M]ain/Off Switch, [Q]uit, [?]Help', 1);
    io.cr();
    const choice = await io.lettersPrompt('Your choice?', 'WASGMQ?');
    io.cr();
    switch (choice) {
      case 'W': session.setNext(new BuyWeaponState()); return;
      case 'A': session.setNext(new BuyArmorState()); return;
      case 'S': session.setNext(new SellWeaponState()); return;
      case 'G': session.setNext(new SellArmorState()); return;
      case 'M':
        io.println('You switch your weapons.', 1);
        io.cr();
        [c.weapons[0], c.weapons[1]] = [c.weapons[1], c.weapons[0]];
        saveCharacter(c);
        session.setNext(new ArmoryPromptState());
        return;
      case 'Q':
        c.location = WhereType.TheTown;
        session.setNext(new TownState());
        return;
      case '?': session.setNext(new ArmoryState()); return;
    }
    session.setNext(new ArmoryPromptState());
  }
}

export class BuyWeaponState {
  async enter(session) {
    const c = session.character;
    const { io } = session;
    const weapons = await loadWeapons();
    if (!weapons.length) {
      io.println('No weapons available!', 6);
      await io.pausePrompt();
      session.setNext(new ArmoryPromptState());
      return;
    }
    io.cr();
    io.println('=--=-- Available Weapons ---=--=', 3);
    io.cr();
    weapons.forEach((w, i) => {
      const affordable = w.cost <= c.coinsHand ? '*' : ' ';
      const rangeStr = w.range > 0 ? ` Range:${w.range}` : '';
      const idx = String(i + 1).padStart(2);
      const name = (w.name || '').padEnd(18);
      const strike = String(w.strike).padEnd(3);
      const cost = String(w.cost).padEnd(4);
      io.println(` ${affordable}${idx}) ${name} Strike:${strike}${rangeStr}  Cost:${cost}`, 1);
    });
    io.cr();
    io.println('  (* = you can afford)', 4);
    io.cr();
    const num = await io.numbersPrompt('Choose Your Weapon (0=Quit):', 0, weapons.length);
    if (num === 0) {
      session.setNext(new ArmoryPromptState());
      return;
    }
    const weapon = weapons[num - 1];
    if (!await io.yesNoQuestion(`Buy ${weapon.name} for ${weapon.cost} coins?`)) {
      io.println("FINE, I didn't want your stupid weapon anyway...", 1);
      io.cr();
      session.setNext(new ArmoryPromptState());
      return;
    }
    if (c.coinsHand < weapon.cost) {
      io.cr();
      io.println("Sorry, I'm not giving these away. Come back when you get more money.", 6);
      io.cr();
      await io.pausePrompt('--More--');
      session.setNext(new ArmoryPromptState());
      return;
    }
    io.cr();
    io.println(`  Weapon 1: ${c.weapons[0].name || 'Empty'}`, 1);
    io.println(`  Weapon 2: ${c.weapons[1].name || 'Empty'}`, 1);
    io.cr();
    const slot = await io.numbersPrompt('Which weapon slot to replace? (1 or 2):', 1, 2);
    c.weapons[slot - 1] = { ...weapon };
    c.coinsHand -= weapon.cost;
    saveCharacter(c);
    io.cr();
    io.println(`Okay, you buy the spiffy ${weapon.name}.`, 3);
    await io.pausePrompt('--More--');
    session.setNext(new ArmoryPromptState());
  }
}

export class BuyArmorState {
  async enter(session) {
    const c = session.character;
    const { io } = session;
    const armors = await loadArmor();
    if (!armors.length) {
      io.println('No armor available!', 6);
      await io.pausePrompt();
      session.setNext(new ArmoryPromptState());
      return;
    }
    io.cr();
    io.println('=--=-- Available Armor ---=--=', 3);
    io.cr();
    armors.forEach((a, i) => {
      const affordable = a.cost <= c.coinsHand ? '*' : ' ';
      const cn = ARMOR_CLASS_NAMES[a.class] ?? 'Body';
      const idx = String(i + 1).padStart(2);
      const name = (a.name || '').padEnd(18);
      const def = String(a.defense).padEnd(3);
      const type = cn.padEnd(9);
      const cost = String(a.cost).padEnd(4);
      io.println(` ${affordable}${idx}) ${name} Def:${def} Type:${type} Cost:${cost}`, 1);
    });
    io.cr();
    io.println('  (* = you can afford)', 4);
    io.cr();
    const num = await io.numbersPrompt('Choose Your Armor (0=Quit):', 0, armors.length);
    if (num === 0) {
      session.setNext(new ArmoryPromptState());
      return;
    }
    const armor = armors[num - 1];
    if (!await io.yesNoQuestion(`Buy ${armor.name} for ${armor.cost} coins?`)) {
      io.println('Okay, your loss. Have fun dying.', 1);
      io.cr();
      session.setNext(new ArmoryPromptState());
      return;
    }
    if (c.coinsHand < armor.cost) {
      io.cr();
      io.println("Sorry, I'm not giving these away. Come back when you get more money.", 6);
      io.cr();
      session.setNext(new ArmoryPromptState());
      return;
    }
    c.armor[0] = { ...armor };
    c.coinsHand -= armor.cost;
    saveCharacter(c);
    io.cr();
    io.println(`Okay, you buy the spiffy ${armor.name}.`, 3);
    await io.pausePrompt('--More--');
    session.setNext(new ArmoryPromptState());
  }
}

export class SellWeaponState {
  async enter(session) {
    const c = session.character;
    const { io } = session;
    const w = c.weapons[0];
    const sellPrice = Math.floor((w.cost ?? 0) / 2);
    if (!w.name || w.name === 'Hands') {
      io.cr();
      io.println("You can't sell your hands. Idiot...", 6);
      io.cr();
      await io.pausePrompt('-More-');
      session.setNext(new ArmoryPromptState());
      return;
    }
    if ((w.cost ?? 0) <= 0) {
      io.cr();
      io.println("Sorry, I don't buy junk.", 6);
      io.cr();
      await io.pausePrompt('-More-');
      session.setNext(new ArmoryPromptState());
      return;
    }
    if (!await io.yesNoQuestion(`Sell your ${w.name} for ${sellPrice} coins?`)) {
      io.cr();
      io.println("Okay, you jerk, I didn't want it anyway.", 3);
      io.cr();
      await io.pausePrompt('-More-');
      session.setNext(new ArmoryPromptState());
      return;
    }
    c.coinsHand += sellPrice;
    c.weapons[0] = { name: 'Hands', strike: 1, range: 0, actionStr: 'punch', cost: 0 };
    saveCharacter(c);
    io.cr();
    io.println('Okay, thanks. Come back anytime.', 1);
    io.cr();
    await io.pausePrompt('-More-');
    session.setNext(new ArmoryPromptState());
  }
}

export class SellArmorState {
  async enter(session) {
    const c = session.character;
    const { io } = session;
    const a = c.armor[0];
    const sellPrice = Math.floor((a.cost ?? 0) / 2);
    if (!a.name || (a.cost ?? 0) <= 0) {
      io.cr();
      io.println("Sorry, I don't buy junk.", 6);
      io.cr();
      await io.pausePrompt('-More-');
      session.setNext(new ArmoryPromptState());
      return;
    }
    if (!await io.yesNoQuestion(`Sell your ${a.name} for ${sellPrice} coins?`)) {
      io.cr();
      io.println("Okay, you jerk, I didn't want it anyway.", 3);
      io.cr();
      await io.pausePrompt('-More-');
      session.setNext(new ArmoryPromptState());
      return;
    }
    c.coinsHand += sellPrice;
    c.armor[0] = { name: '', defense: 0, cost: 0 };
    saveCharacter(c);
    io.cr();
    io.println('Okay, thanks. Come back anytime.', 1);
    io.cr();
    await io.pausePrompt('-More-');
    session.setNext(new ArmoryPromptState());
  }
}
